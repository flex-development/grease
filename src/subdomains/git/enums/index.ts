/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
