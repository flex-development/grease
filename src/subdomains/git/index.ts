/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './options'
export * from './providers'
export * from './queries'
export type * from './types'
