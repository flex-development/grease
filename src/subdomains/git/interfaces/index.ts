/**
 * @file Entry Point - Interfaces
 * @module grease/git/interfaces
 */

export type { default as ICommit } from './commit.interface'
