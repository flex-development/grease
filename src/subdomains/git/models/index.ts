/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
