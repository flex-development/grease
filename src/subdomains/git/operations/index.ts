/**
 * @file Entry Point - Operations
 * @module grease/git/operations
 */

export { default as TagOperationHandler } from './tag.handler'
export { default as TagOperation, type TagOperationDTO } from './tag.operation'
