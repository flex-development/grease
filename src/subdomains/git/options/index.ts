/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitOptions } from './git.options'
