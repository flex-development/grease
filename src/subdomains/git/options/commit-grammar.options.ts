/**
 * @file Options - CommitGrammarOptions
 * @module grease/git/options/CommitGrammarOptions
 */

import { RepositoryProvider } from '#src/git/enums'
import {
  get,
  isString,
  select,
  sort,
  unique,
  type Partial
} from '@flex-development/tutils'
import {
  IsArray,
  IsEnum,
  IsString,
  MinLength
} from 'class-validator'

/**
 * Commit parser grammar options.
 *
 * @class
 */
class CommitGrammarOptions {
  /**
   * Reference action keywords.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} actions
   */
  @IsArray()
  @IsString({ each: true })
  @MinLength(1, { each: true })
  public actions: string[]

  /**
   * Issue reference prefixes.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} issues
   */
  @IsArray()
  @IsString({ each: true })
  public issues: string[]

  /**
   * Pull request reference prefixes.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} pr
   */
  @IsArray()
  @IsString({ each: true })
  public pr: string[]

  /**
   * Repository provider name.
   *
   * @default RepositoryProvider.GITHUB
   *
   * @public
   * @instance
   * @member {RepositoryProvider} provider
   */
  @IsEnum(RepositoryProvider)
  public provider: RepositoryProvider

  /**
   * Tag prefix.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} tagprefix
   */
  @IsString()
  public tagprefix: string

  /**
   * Create a new options object.
   *
   * @param {Partial<CommitGrammarOptions>?} [opts] - Commit grammar options
   */
  constructor(opts?: Partial<CommitGrammarOptions>) {
    this.actions = select(get(opts, 'actions', []), isString)
    this.issues = select(get(opts, 'issues', []), isString)
    this.pr = select(get(opts, 'pr', []), isString)
    this.provider = get(opts, 'provider', RepositoryProvider.GITHUB)
    this.tagprefix = get(opts, 'tagprefix', '')

    // add common provider actions and prefixes
    this.actions.push('close', 'closed', 'closes', 'fix', 'fixed')
    this.issues.push('#')
    this.provider !== RepositoryProvider.GITLAB && this.pr.push('#')

    // add actions and prefixes based on provider
    switch (this.provider) {
      case RepositoryProvider.BITBUCKET:
        this.actions.push(
          'closing',
          'fixes',
          'fixing',
          'resolve',
          'resolved',
          'resolves',
          'resolving'
        )
        this.issues.unshift('JRA-', 'JIRA-')
        break
      case RepositoryProvider.GITLAB:
        this.actions.push('closing', 'fixes', 'fixing')
        this.pr.push('!')
        break
      default:
        this.actions.push('fixes', 'resolve', 'resolved', 'resolves')
        this.issues.push('GH-')
        break
    }

    // ensure actions and prefixes are unique
    this.actions = sort(unique(this.actions), (a, b) => b.length - a.length)
    this.issues = unique(this.issues)
    this.pr = unique(this.pr)
  }
}

export default CommitGrammarOptions
