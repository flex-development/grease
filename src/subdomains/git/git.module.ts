/**
 * @file GitModule
 * @module grease/git/GitModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Global, Module } from '@nestjs/common'
import { CqrsModule } from '@nestjs/cqrs'
import { TagOperationHandler } from './operations'
import { GitService } from './providers'
import { CommitQueryHandler, TagQueryHandler } from './queries'

/**
 * Git operations module.
 *
 * @class
 */
@Global()
@Module({
  exports: [GitService],
  imports: [CqrsModule],
  providers: [
    CommitQueryHandler,
    GitService,
    LoggerService,
    TagOperationHandler,
    TagQueryHandler,
    ValidationService
  ]
})
class GitModule {}

export default GitModule
