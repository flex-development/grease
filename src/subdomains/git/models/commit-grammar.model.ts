/**
 * @file Models - CommitGrammar
 * @module grease/git/models/CommitGrammar
 */

import { CommitGrammarOptions } from '#src/git/options'
import {
  join,
  regexp,
  select,
  template,
  unique,
  type Partial
} from '@flex-development/tutils'

/**
 * Commit parser grammar.
 *
 * @class
 */
class CommitGrammar {
  /**
   * Grammar options.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitGrammarOptions} options
   */
  protected readonly options: CommitGrammarOptions

  /**
   * Create a new commit grammar instance.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @param {Partial<CommitGrammarOptions>?} [options] - Grammar options
   */
  constructor(options?: Partial<CommitGrammarOptions>) {
    this.options = new CommitGrammarOptions(options)
  }

  /**
   * Regular expression matching fields in a raw commit.
   *
   * @see https://regex101.com/r/QnFMsV
   *
   * @public
   *
   * @return {RegExp} Raw commit field regex
   */
  public get field(): RegExp {
    return /^-(?<field>[^\s$]+?)-\n(?<value>.*?(?=\n*^-\S+?-)|(?:.*$))/gms
  }

  /**
   * Regular expression matching a commit header.
   *
   * @see https://regex101.com/r/WAJeLp
   * @see https://conventionalcommits.org
   *
   * @public
   *
   * @return {RegExp} Commit header regex
   */
  public get header(): RegExp {
    return /^(?<type>[a-z]+)(?:\((?<scope>[a-z-]+)\))?(?<breaking>!)?: +(?<subject>(?:.+ \([!#](?<pr>\d+)\))|.+)/i
  }

  /**
   * Regular expression matching mentions (e.g. `@unicornware`) in a raw commit.
   *
   * @see https://regex101.com/r/upbRpj
   *
   * @public
   *
   * @return {RegExp} Mention regex
   */
  public get mention(): RegExp {
    return /\B(?<mention>@(?<user>[\w-]{1,38}(?=[^\w-]\B|\s)))/g
  }

  /**
   * Get a regular expression matching references in a raw commit.
   *
   * @see https://regex101.com/r/Thsp1M
   *
   * @public
   *
   * @return {RegExp} Reference regex
   */
  public get reference(): RegExp {
    /**
     * Pattern matching issue references in commit bodies.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '(?:(?:(?<action>{actions}) +)|(?<repository>(?<owner>[\\da-z](?:-(?=[\\da-z])|[\\da-z]){0,38}(?<=[\\da-z]))\\/(?<repo>\\S+)))?(?<ref>(?<prefix>{prefixes})(?<number>\\d+))'

    return new RegExp(
      template(pattern, {
        actions: join(this.options.actions, '|'),
        prefixes: join(select(unique([
          ...this.options.issues,
          ...this.options.pr
        ]), null, regexp), '|')
      }),
      'gi'
    )
  }

  /**
   * Regular expression matching a semantic release tag.
   *
   * @see https://regex101.com/r/WyQyHZ
   * @see https://semver.org
   *
   * @public
   *
   * @return {RegExp} Semantic release tag regex
   */
  public get tag(): RegExp {
    /**
     * Pattern matching semantic release tags.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '^(?<prefix>{prefix})(?<version>(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<prerelease>(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*))*))?(?:\\+(?<metadata>[\\dA-Za-z-]+(?:\\.[\\dA-Za-z-]+)*))?)$'

    return new RegExp(template(pattern, { prefix: this.options.tagprefix }))
  }

  /**
   * Regular expression matching trailers in a raw commit.
   *
   * @see https://git-scm.com/docs/git-interpret-trailers
   * @see https://regex101.com/r/I56Kgg
   *
   * @return {RegExp} Git trailer regex
   */
  public get trailer(): RegExp {
    return /(?<=^|\n)(?<token>[\w -]+?(?=:)): (?<value>[\S\s]+?(?:(?=\n[\w -]+?(?=:))|(?=\n*$)))/g
  }
}

export default CommitGrammar
