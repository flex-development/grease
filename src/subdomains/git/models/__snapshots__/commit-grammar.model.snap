// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:git/models/CommitGrammar > #field > should match raw commit fields 1`] = `
[
  {
    "0": "-author.email-
unicornware@flexdevelopment.llc",
    "1": "author.email",
    "2": "unicornware@flexdevelopment.llc",
    "groups": {
      "field": "author.email",
      "value": "unicornware@flexdevelopment.llc",
    },
  },
  {
    "0": "-author.name-
Lexus Drumgold",
    "1": "author.name",
    "2": "Lexus Drumgold",
    "groups": {
      "field": "author.name",
      "value": "Lexus Drumgold",
    },
  },
  {
    "0": "-body-
",
    "1": "body",
    "2": "",
    "groups": {
      "field": "body",
      "value": "",
    },
  },
  {
    "0": "-date-
2021-09-18T23:29:18-04:00",
    "1": "date",
    "2": "2021-09-18T23:29:18-04:00",
    "groups": {
      "field": "date",
      "value": "2021-09-18T23:29:18-04:00",
    },
  },
  {
    "0": "-hash-
e003836",
    "1": "hash",
    "2": "e003836",
    "groups": {
      "field": "hash",
      "value": "e003836",
    },
  },
  {
    "0": "-header-
release: @flex-development/grease@2.0.0",
    "1": "header",
    "2": "release: @flex-development/grease@2.0.0",
    "groups": {
      "field": "header",
      "value": "release: @flex-development/grease@2.0.0",
    },
  },
  {
    "0": "-sha-
e00383628638335e729d1cad2971b44ba03d3074",
    "1": "sha",
    "2": "e00383628638335e729d1cad2971b44ba03d3074",
    "groups": {
      "field": "sha",
      "value": "e00383628638335e729d1cad2971b44ba03d3074",
    },
  },
  {
    "0": "-tags-
tag: grease@2.0.0",
    "1": "tags",
    "2": "tag: grease@2.0.0",
    "groups": {
      "field": "tags",
      "value": "tag: grease@2.0.0",
    },
  },
  {
    "0": "-trailers-
",
    "1": "trailers",
    "2": "",
    "groups": {
      "field": "trailers",
      "value": "",
    },
  },
]
`;

exports[`unit:git/models/CommitGrammar > #header > should match raw commit header with [type,breaking,subject] 1`] = `
{
  "0": "feat!: watch mode",
  "1": "feat",
  "2": undefined,
  "3": "!",
  "4": "watch mode",
  "groups": {
    "breaking": "!",
    "scope": undefined,
    "subject": "watch mode",
    "type": "feat",
  },
}
`;

exports[`unit:git/models/CommitGrammar > #header > should match raw commit header with [type,scope,breaking,subject] 1`] = `
{
  "0": "refactor(utils)!: [\`analyzeOutputs\`] intake \`esbuild.Metafile.outputs\`",
  "1": "refactor",
  "2": "utils",
  "3": "!",
  "4": "[\`analyzeOutputs\`] intake \`esbuild.Metafile.outputs\`",
  "groups": {
    "breaking": "!",
    "scope": "utils",
    "subject": "[\`analyzeOutputs\`] intake \`esbuild.Metafile.outputs\`",
    "type": "refactor",
  },
}
`;

exports[`unit:git/models/CommitGrammar > #header > should match raw commit header with [type,scope,subject] 1`] = `
{
  "0": "fix(config): prevent \`import.meta.url\` from being rewritten",
  "1": "fix",
  "2": "config",
  "3": undefined,
  "4": "prevent \`import.meta.url\` from being rewritten",
  "groups": {
    "breaking": undefined,
    "scope": "config",
    "subject": "prevent \`import.meta.url\` from being rewritten",
    "type": "fix",
  },
}
`;

exports[`unit:git/models/CommitGrammar > #header > should match raw commit header with [type,subject] 1`] = `
{
  "0": "ci: add @dependabot config",
  "1": "ci",
  "2": undefined,
  "3": undefined,
  "4": "add @dependabot config",
  "groups": {
    "breaking": undefined,
    "scope": undefined,
    "subject": "add @dependabot config",
    "type": "ci",
  },
}
`;

exports[`unit:git/models/CommitGrammar > #mention > should match mentions in raw commit 1`] = `
[
  {
    "0": "@dependabot",
    "1": "@dependabot",
    "2": "dependabot",
    "groups": {
      "mention": "@dependabot",
      "user": "dependabot",
    },
  },
]
`;

exports[`unit:git/models/CommitGrammar > #reference > should match references in raw commit 1`] = `
[
  {
    "0": "import-js/eslint-plugin-import#2843",
    "1": undefined,
    "2": "import-js/eslint-plugin-import",
    "3": "import-js",
    "4": "eslint-plugin-import",
    "5": "#2843",
    "6": "#",
    "7": "2843",
    "groups": {
      "action": undefined,
      "number": "2843",
      "owner": "import-js",
      "prefix": "#",
      "ref": "#2843",
      "repo": "eslint-plugin-import",
      "repository": "import-js/eslint-plugin-import",
    },
  },
]
`;

exports[`unit:git/models/CommitGrammar > #tag > should match semantic release tag 1`] = `
{
  "0": "grease@2.0.0+ae99e7e800dcbef9e7b5720493e073e7dc7827ab",
  "1": "grease@",
  "2": "2.0.0+ae99e7e800dcbef9e7b5720493e073e7dc7827ab",
  "3": "2",
  "4": "0",
  "5": "0",
  "6": undefined,
  "7": "ae99e7e800dcbef9e7b5720493e073e7dc7827ab",
  "groups": {
    "major": "2",
    "metadata": "ae99e7e800dcbef9e7b5720493e073e7dc7827ab",
    "minor": "0",
    "patch": "0",
    "prefix": "grease@",
    "prerelease": undefined,
    "version": "2.0.0+ae99e7e800dcbef9e7b5720493e073e7dc7827ab",
  },
}
`;

exports[`unit:git/models/CommitGrammar > #tag > should match semantic version 1`] = `
{
  "0": "3.0.0-alpha.1",
  "1": "",
  "2": "3.0.0-alpha.1",
  "3": "3",
  "4": "0",
  "5": "0",
  "6": "alpha.1",
  "7": undefined,
  "groups": {
    "major": "3",
    "metadata": undefined,
    "minor": "0",
    "patch": "0",
    "prefix": "",
    "prerelease": "alpha.1",
    "version": "3.0.0-alpha.1",
  },
}
`;

exports[`unit:git/models/CommitGrammar > #trailer > should match trailers in raw commit 1`] = `
[
  {
    "0": "token: This is a very long value, with spaces and
  newlines in it.",
    "1": "token",
    "2": "This is a very long value, with spaces and
  newlines in it.",
    "groups": {
      "token": "token",
      "value": "This is a very long value, with spaces and
  newlines in it.",
    },
  },
  {
    "0": "BREAKING CHANGE: \`extends\` key now used for extending other configs",
    "1": "BREAKING CHANGE",
    "2": "\`extends\` key now used for extending other configs",
    "groups": {
      "token": "BREAKING CHANGE",
      "value": "\`extends\` key now used for extending other configs",
    },
  },
  {
    "0": "BREAKING-CHANGE: use JavaScript features not available in Node 6.",
    "1": "BREAKING-CHANGE",
    "2": "use JavaScript features not available in Node 6.",
    "groups": {
      "token": "BREAKING-CHANGE",
      "value": "use JavaScript features not available in Node 6.",
    },
  },
  {
    "0": "Reviewed-by: Z",
    "1": "Reviewed-by",
    "2": "Z",
    "groups": {
      "token": "Reviewed-by",
      "value": "Z",
    },
  },
  {
    "0": "Refs: #123",
    "1": "Refs",
    "2": "#123",
    "groups": {
      "token": "Refs",
      "value": "#123",
    },
  },
]
`;
